File: boj/DataStructure/BOJ_1717.cpp
/*
 백준 1717 집합의 표현
 21.07.12
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;

#define INF 1e9
#define endl "\n"

typedef long long ll;
typedef double dd;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;

int n, m, parent[1000001];

int findParent(int a) {
    if (parent[a] == a) return a;
    return parent[a] = findParent(parent[a]);
}

void unionParent(int a, int b) {
    int x = findParent(parent[a]);
    int y = findParent(parent[b]);

    if (x < y) parent[y] = x;
    else parent[x] = y;
}

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);

    cin >> n >> m;

    for (int i = 0; i <= n; ++i) {
        parent[i] = i;
    }

    for (int i = 0; i < m; ++i) {
        int a, b, c;
        cin >> a >> b >> c;

        if (a == 0) unionParent(b, c);
        else {
            if (findParent(b) != findParent(c)) {
                cout << "NO\n";
            } else {
                cout << "YES\n";
            }
        }
    }

    return 0;
}

File: boj/DataStructure/BOJ_1717.swift
/*
 백준 1717 집합의 표현
 21.08.21
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0]
var m = input[1]

var parent = [Int](repeating: -1, count: n + 1)

for i in parent.indices {
  parent[i] = i
}

func findParent(_ idx: Int) -> Int {
  if idx != parent[idx] {
    parent[idx] = findParent(parent[idx])
  }

  return parent[idx]
}

func unionParent(_ a: Int, _ b: Int) {
  let x = findParent(a)
  let y = findParent(b)

  if x < y {
    parent[y] = x
  } else {
    parent[x] = y
  }
}

while m > 0 {
  m -= 1

  let query = readLine()!.split(separator: " ").map { Int(String($0))! }

  if query[0] == 0 {
    unionParent(query[1], query[2])
  } else {
    if findParent(query[1]) == findParent(query[2]) {
      print("YES")
    } else {
      print("NO")
    }
  }
}